File: js/parse_composite.js
let parse_composite = function(text) {
    let lines = text.split("\n"),
        xmin, xmax, sense, anti, xmin_curr, xmax_curr, offset,
        i = 0;
    while (i < lines.length) {
        let line = lines[i];
        // Skip empty lines
        if (line.trim() === "") {
            i++;
            continue
        };
        // Tab-separated fields
        let fields = line.split("\t");
        // If the first field is empty or "NAME" it is the x domain
        if (fields[0] === "" || fields[0] === "NAME") {
            xmin_curr = parseInt(fields[1]);
            xmax_curr = parseInt(fields[fields.length - 1]);
            // If the x domain starts at 0 shift it to the left
            if (xmin_curr === 0) {
                xmin_curr -= Math.floor(xmax_curr / 2)
                xmax_curr -= Math.floor(xmax_curr / 2)
            };
            // If the x domain is not defined yet, define it
            if (xmin === undefined || xmax === undefined) {
                xmin = xmin_curr;
                xmax = xmax_curr;
                sense = Array(xmax_curr - xmin_curr + 1).fill(0);
                anti = Array(xmax_curr - xmin_curr + 1).fill(0)
            // If the x domain is defined, extend it if necessary
            } else {
                if (xmin_curr < xmin) {
                    sense.unshift(Array(xmin - xmin_curr).fill(0));
                    anti.unshift(Array(xmin - xmin_curr).fill(0));
                    xmin = xmin_curr
                };
                if (xmax_curr > xmax) {
                    sense.push(Array(xmax_curr - xmax).fill(0));
                    anti.push(Array(xmax_curr - xmax).fill(0));
                    xmax = xmax_curr
                }
            };
            // Calculate the offset of the current x domain
            offset = xmin_curr - xmin
        // Add the values to the sense and anti arrays
        } else if (fields[0].toLowerCase().includes("sense")) {
            for (let j = 1; j < fields.length; j++) {
                sense[offset + j - 1] += parseFloat(fields[j])
            }
        } else if (fields[0].toLowerCase().includes("anti")) {
            for (let j = 1; j < fields.length; j++) {
                anti[offset + j - 1] += parseFloat(fields[j])
            }
        // If the first field is not empty or "NAME" and does not contain "sense" or "anti" it is combined or midpoint data
        } else {
            for (let j = 1; j < fields.length; j++) {
                sense[offset + j - 1] += parseFloat(fields[j]) / 2;
                anti[offset + j - 1] += parseFloat(fields[j]) / 2
            }
        };

        i++
    };

    return {xmin: xmin, xmax: xmax, sense: sense, anti: anti}
};

let get_prefixes_from_multiple_composite = function(text) {
    let lines = text.split("\n"),
        names_list = [],
        i = 0;

    while (i < lines.length) {
        let line = lines[i];
        // Skip empty lines
        if (line.trim() === "") {
            i++;
            continue
        };
        // Get the first field
        let col0 = line.split("\t")[0];
        if (col0 === "" || col0 === "NAME") {
            // Get the names of the composites for lines immediately following the xdomain
            names_list.push(lines[++i].split("\t")[0])
        }

        i++
    };

    // Take the first name and split it by "_"
    let split_name = names_list[0].split("_"),
        idx;
    // Iterate over each possible prefix-suffix split
    for (let i = 1; i < split_name.length - 1; i++) {
        let prefix = split_name.slice(0, i).join("_"),
            suffix = split_name.slice(i).join("_"),
            n_prefix = names_list.reduce((a, c) => a + c.startsWith(prefix), 0),
            n_suffix = names_list.reduce((a, c) => a + c.endsWith(suffix), 0);
        if (n_prefix * n_suffix === names_list.length) {
            if (n_suffix === names_list.length) {
                idx = idx === undefined ? i : idx;
                break
            };
            idx = i
        }
    };

    let suffix = split_name.slice(idx).join("_");
    // Get the prefixes by removing the suffix from the names
    return names_list.filter(n => n.endsWith(suffix)).map(n => n.slice(0, -suffix.length))
};

let parse_multiple_composite = function(text, prefix) {
    let lines = text.split("\n"),
        composites = {},
        xmin, xmax, id, sense, anti,
        i = 0,
        save_comp = false;

    while (i < lines.length) {
        let line = lines[i];
        // Skip empty lines
        if (line.trim() === "") {
            i++;
            continue
        };
        // Get the first field
        let fields = line.split("\t"),
            col0 = fields[0];
        if (col0 === "" || col0 === "NAME") {
            // If the x domain is defined, save the composite
            if (save_comp) {
                console.log(id)
                composites[id] = {xmin: xmin, xmax: xmax, sense: sense, anti: anti}
            };
            save_comp = false;
            // Get the new x domain
            xmin = parseInt(fields[1]);
            xmax = parseInt(fields[fields.length - 1]);
            // If the x domain starts at 0 shift it to the left
            if (xmin === 0) {
                xmin -= Math.floor(xmax / 2)
                xmin -= Math.floor(xmax / 2)
            }
        } else if (col0.startsWith(prefix)){
            id = col0.slice(prefix.length).split("_")[0];
            console.log(id)
            save_comp = true;
            if (col0.toLowerCase().includes("sense")) {
                sense = fields.slice(1).map(parseFloat)
            } else if (col0.toLowerCase().includes("anti")) {
                anti = fields.slice(1).map(parseFloat)
            } else {
                sense = fields.slice(1).map(parseFloat).map(x => x / 2);
                anti = fields.slice(1).map(parseFloat).map(x => x / 2)
            }
        };

        i++
    };

    // Save the last composite
    if (save_comp) {
        composites[id] = {xmin: xmin, xmax: xmax, sense: sense, anti: anti}
    };
    console.log(composites)
    return composites
}

File: python/parseComposite.py
import csv
import xml.dom.minidom as dom
import argparse
import math
import composite
import os

"""
This script is designed to parse data from composite files, extract sense and anti-sense data, and organize the data into a
structured format. 

The key functions in the script:
1. parse_simple
2. get_prefixes_from_multiple_composites
3. parse_multiple_composite
4. 
"""

# Returns a simple composite from a single file
def parse_simple(file):
    fileArr = open(file, "r").read().split("\n")
    xmin = None
    xmax = None
    sense = []
    anti = []
    xmin_curr = 0
    xmax_curr = 0
    offset = 0
    for line in fileArr:
        # Skip empty
        if len(line.strip()) == 0 :
            continue
        # Separate fields
        fields = line.split("\t")
        if not fields[0].strip() or fields[0] == "NAME":
            xmin_curr = int(float(fields[1]))
            xmax_curr = int(float(fields[-1]))
            # If the x domain starts at 0 shift it to the left
            if xmin_curr == 0:
                xmin_curr -= math.floor(xmax_curr / 2)
                xmax_curr -= math.floor(xmax_curr / 2)
            # If the x domain is not defined yet, define it
            if xmin == None or xmax == None:
                xmin = xmin_curr
                xmax = xmax_curr
            # Redefine min and max if necessary
            xmax = max(xmax_curr, xmax)
            xmin = min(xmin_curr, xmin)
            sense = [0] * (xmax - xmin + 1)
            anti = [0] * (xmax - xmin + 1)
        # Add the values to sense and anti arrays
        if "sense" in fields[0].lower():
            i = 1
            while i < len(fields):
                sense[offset + i - 1] += float(fields[i])
                i += 1
        elif "anti" in fields[0].lower():
            i = 1
            while i < len(fields):
                anti[offset + i - 1] += float(fields[i])
                i += 1
        # If the first field is not empty or "NAME" and does not contain "sense" or "anti" parse as combined or midpoint data
        elif not (fields[0] == "" or fields[0] == "NAME"):
            i = 1
            while i < len(fields):
                sense[offset + i - 1] += float(fields[i]) / 2
                anti[offset + i - 1] += float(fields[i]) / 2
    return composite.SimpleComposite(xmin, xmax, sense, anti, os.path.basename(file).split('_')[0])

# Returns list of prefixes from multi-composite file, mimics the plotter method
def get_prefixes_from_multiple_composites(file):
    lines = open(file, "r").read().split("\n")
    names_list = []
    i = 0
    while i < len(lines):
        line = lines[i]
        # Skip empty 
        if line.strip() == "":
            i += 1
            continue
        # Get the first field
        col0 = line.split("\t")[0]
        if col0 == "" or col0[0] == "NAME":
            # Get the names of the composites for lines immediately following the xdomain
            i += 1
            names_list.append(lines[i].split("\t")[0])
        i += 1
    # Take the first name and split it by "_"
    split_name = names_list[0].split("_")
    idx = None
    # Iterate over each possible prefix-suffix split
    for i in range(1, len(split_name) - 1):
        prefix = "_".join(split_name[:i])
        suffix = "_".join(split_name[i:])
        n_prefix = sum(1 for n in names_list if n.startswith(prefix))
        n_suffix = sum(1 for n in names_list if n.endswith(suffix))
        if n_prefix * n_suffix == len(names_list):
            if n_suffix == len(names_list):
                idx = i if idx is None else idx
                break
            idx = i
    suffix = "_".join(split_name[idx:])
    # Get the prefixes by removing the suffix from the names
    return [n[:-len(suffix)] for n in names_list if n.endswith(suffix)]

# Returns dictionary with composite from multi-composite file, mimics the plotter method
def parse_multiple_composite(file, prefix):
    lines = open(file, "r").read().split("\n")
    composites = {}
    xmin = None
    xmax = None
    sense = []
    anti = []
    i = 0
    id = 0
    save_comp = False
    while i < len(lines):
        line = lines[i]
        # Skip empty
        if line.strip() == "":
            i += 1
            continue
        # Get the first field
        fields = line.split("\t")
        col0 = fields[0]
        if not col0.strip() or col0 == "NAME":
            # If the x domain is defined, save the composite
            if save_comp:
                composites[id] = composite.SimpleComposite(xmin, xmax, sense, anti, id)
            save_comp = False
            # Get the nex x domain
            fields = [field for field in fields if field.strip()]
            xmin = int(float(fields[0]))
            xmax = int(float(fields[-1]))
            # If the x domain starts at 0 shift it to the left
            if xmin == 0:
                xmin -= math.floor(xmax / 2)
                xmin -= math.floor(xmax / 2)
        elif col0.startswith(prefix):
            id = col0[len(prefix):].split("_")[0]
            save_comp = True
            # Add the values to sense and anti arrays
            fields = [field for field in fields if field.strip()]
            if "sense" in fields[0].lower():
                sense = [float(val) for val in fields[1:]]
            elif "anti" in fields[0].lower():
                anti = [float(val) for val in fields[1:]]
            else:
                sense = [float(val) / 2 for val in fields[1:]]
                anti = [float(val) / 2 for val in fields[1:]]
        i += 1
    # Save the last composite
    if save_comp:
        composites[id] = composite.SimpleComposite(xmin, xmax, sense, anti, id)
    return composites